File: llm_proxy/routing.py
"""Pure helpers for request model routing and model-name rewrites."""

from __future__ import annotations

import json
from collections.abc import Sequence
from dataclasses import dataclass

from .models import RuntimeTarget


@dataclass(frozen=True)
class TargetSelection:
    target: RuntimeTarget
    request_model: str | None
    upstream_model: str | None


def request_model_from_body(request_body: bytes) -> str | None:
    try:
        loaded = json.loads(request_body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return None
    if not isinstance(loaded, dict):
        return None
    model = loaded.get("model")
    return model if isinstance(model, str) else None


def select_target_by_model(
    targets: Sequence[RuntimeTarget],
    default_target_id: str,
    request_body: bytes,
) -> TargetSelection:
    """Select a target using the top-level request model field."""
    if not targets:
        raise ValueError("ProxyServer config must include at least one target.")
    default_target = next((target for target in targets if str(target.get("id")) == default_target_id), targets[0])
    request_model = request_model_from_body(request_body)
    if request_model:
        for target in targets:
            if target is not default_target and not bool(target.get("enabled", True)):
                continue
            mappings = target.get("model_mappings")
            if not isinstance(mappings, list):
                continue
            for mapping in mappings:
                if not isinstance(mapping, dict):
                    continue
                listen_model = mapping.get("listen")
                if listen_model == request_model:
                    upstream_model = mapping.get("upstream")
                    return TargetSelection(
                        target=target,
                        request_model=request_model,
                        upstream_model=upstream_model if isinstance(upstream_model, str) and upstream_model else None,
                    )
    return TargetSelection(target=default_target, request_model=request_model, upstream_model=None)


def rewrite_request_model(request_body: bytes, upstream_model: str | None) -> bytes:
    if not upstream_model:
        return request_body
    try:
        loaded = json.loads(request_body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return request_body
    if not isinstance(loaded, dict):
        return request_body
    loaded["model"] = upstream_model
    return json.dumps(loaded, ensure_ascii=False, separators=(",", ":")).encode("utf-8")

File: src/proxy/routing.ts
import { isRecord } from "../shared/index.js";

const UTF8_DECODER = new TextDecoder("utf-8", { fatal: true });
const UTF8_ENCODER = new TextEncoder();

export interface RoutingTarget {
  readonly id: string;
  readonly enabled: boolean;
  readonly model_mappings: readonly {
    readonly listen: string;
    readonly upstream: string;
  }[];
}

export interface TargetSelection<TTarget extends RoutingTarget = RoutingTarget> {
  readonly target: TTarget;
  readonly requestModel: string | undefined;
  readonly upstreamModel: string | undefined;
}

export function requestModelFromBody(requestBody: Uint8Array): string | undefined {
  let loaded: unknown;
  try {
    loaded = JSON.parse(UTF8_DECODER.decode(requestBody));
  } catch {
    return undefined;
  }
  if (!isRecord(loaded)) {
    return undefined;
  }
  const model = loaded["model"];
  return typeof model === "string" ? model : undefined;
}

export function selectTargetByModel<TTarget extends RoutingTarget>(
  targets: readonly TTarget[],
  defaultTargetId: string,
  requestBody: Uint8Array,
): TargetSelection<TTarget> {
  const firstTarget = targets[0];
  if (firstTarget === undefined) {
    throw new TypeError("ProxyServer config must include at least one target.");
  }
  const defaultTarget = targets.find(({ id }) => id === defaultTargetId) ?? firstTarget;
  const requestModel = requestModelFromBody(requestBody);
  if (requestModel !== undefined && requestModel !== "") {
    for (const target of targets) {
      if (target !== defaultTarget && !target.enabled) {
        continue;
      }
      for (const mapping of target.model_mappings) {
        if (modelPatternMatches(mapping.listen, requestModel)) {
          const upstreamModel =
            mapping.listen.includes("*") && mapping.upstream === mapping.listen
              ? requestModel
              : mapping.upstream;
          return {
            target,
            requestModel,
            upstreamModel: upstreamModel === "" ? undefined : upstreamModel,
          };
        }
      }
    }
  }
  return { target: defaultTarget, requestModel, upstreamModel: undefined };
}

export function modelPatternMatches(pattern: string, model: string): boolean {
  if (!pattern.includes("*")) {
    return pattern === model;
  }

  let patternIndex = 0;
  let modelIndex = 0;
  let lastStarIndex = -1;
  let modelIndexAfterStar = -1;

  while (modelIndex < model.length) {
    if (patternIndex < pattern.length && pattern[patternIndex] === model[modelIndex]) {
      patternIndex += 1;
      modelIndex += 1;
      continue;
    }
    if (pattern[patternIndex] === "*") {
      lastStarIndex = patternIndex;
      patternIndex += 1;
      modelIndexAfterStar = modelIndex;
      continue;
    }
    if (lastStarIndex === -1) {
      return false;
    }
    patternIndex = lastStarIndex + 1;
    modelIndexAfterStar += 1;
    modelIndex = modelIndexAfterStar;
  }

  while (pattern[patternIndex] === "*") {
    patternIndex += 1;
  }
  return patternIndex === pattern.length;
}

export function rewriteRequestModel(
  requestBody: Uint8Array,
  upstreamModel: string | undefined,
): Uint8Array {
  if (upstreamModel === undefined || upstreamModel === "") {
    return requestBody;
  }
  let loaded: unknown;
  try {
    loaded = JSON.parse(UTF8_DECODER.decode(requestBody));
  } catch {
    return requestBody;
  }
  if (!isRecord(loaded)) {
    return requestBody;
  }
  loaded["model"] = upstreamModel;
  return UTF8_ENCODER.encode(JSON.stringify(loaded));
}
